### PR TITLE
[STORM-3723] Fix isAnyPosixProcessPidDirAlive for PID reassignment

### PR DESCRIPTION
**Pull request by bipinprasad**
## What is the purpose of the change

*Fix isAnyPosixProcessPidDirAlive to return correct value when all PIDs are reassigned to different user*

## How was the change tested

*Modify ServerUtilsTest to add this new condition*

### CODE REVIEW

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 1]
```diff
@@ -106,7 +116,15 @@ public void testExtractZipFileDisallowsPathTraversalWhenUsingPrefix() throws Exc
                     continue;
                 }
                 try {
-                    pids.add(Long.parseLong(line.split("\\s")[0]));
+                     String pidStr = line.split("\\s")[0];
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:119]: indention seems off

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 2]
```diff
@@ -272,18 +290,28 @@ public void testIsAnyPosixProcessPidDirAliveMockingFileOwnerUid() throws Excepti
...
-            String currentUser = System.getProperty("user.name");
-            boolean status = ServerUtils.isAnyPosixProcessPidDirAlive(pids, currentUser, mockFileOwnerToUid);
+            boolean status = ServerUtils.isAnyPosixProcessPidDirAlive(allPids, currentUser, mockFileOwnerToUid);
             String err = String.format("(mockFileOwnerToUid=%s) Expecting user %s to own at least one process",
                     mockFileOwnerToUid, currentUser);
             assertTrue(err, status);
         }
+
+        // simulate reassignment of all process id to a different user (root)
+        for (boolean mockFileOwnerToUid: Arrays.asList(true, false)) {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:309]: Thanks for adding the unit test

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java [hunk 3]
```diff
@@ -272,18 +290,28 @@ public void testIsAnyPosixProcessPidDirAliveMockingFileOwnerUid() throws Excepti
             LOG.info("Test testIsAnyPosixProcessPidDirAlive is designed to run on systems with /proc directory only, marking as success");
             return;
         }
-        Collection<Long> pids = getRunningProcessIds();
-        assertFalse(pids.isEmpty());
+        Collection<Long> allPids = getRunningProcessIds(null);
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:293]: It doesn't seem very necessary to get all pids. We just need the pids from currentUser (`System.getProperty("user.name")`)

File: storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java [hunk 4]
```diff
@@ -1309,8 +1309,6 @@ public static boolean isAnyPosixProcessPidDirAlive(Collection<Long> pids, String
                         + "likely pid {} was reused for a new process for expectedUser {}",
                         pidDir, actualUser, expectedUser, pid, actualUser);
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java:1310]: Looks like the log should be `a new process for actualUser {}"`

And is it possible to print out the cmdline of this processId so we know what the process really is? It seems unlikely that a processId is reused by another user (or root) so quickly. I hope to get more details when this happens. Thanks

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 3 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:293]: This tests the case where some (but not all) PIDs have been reassigned.

File: storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java (hunk 1 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/utils/ServerUtilsTest.java:119]: fixed

File: storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java (hunk 4 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java:1310]: fixed log

File: storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java [hunk 5]
```diff
@@ -1295,22 +1295,45 @@ public static boolean isAnyPosixProcessPidDirAlive(Collection<Long> pids, String
...
+     *
+     * @param pidDir PID directory (/proc/&lt;pid&gt;)
+     * @return process description string
+     */
+    private static String getProcessDesc(File pidDir) {
+        String comm = "";
+        try {
+            comm = String.join(", ", Files.readAllLines(pidDir.toPath().resolve("comm")));
+        } catch (IOException ex) {
+            // ignore
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/utils/ServerUtils.java:1329]: Maybe log a warning here and below?

**Review by Ethanlm — APPROVED** [review]
Thanks. It looks good to me. If we could add some warnings in getProcessDesc for IOException, it will be nice.

**Review by Ethanlm — APPROVED** [review]
Thanks

**Pull request merged**